File: android-compiler/src/main/java/com/pranav/lib_android/task/java/ECJCompilationTask.java
package com.pranav.lib_android.task.java;

import android.content.Context;
import android.content.SharedPreferences;

import com.pranav.common.util.ConcurrentUtil;
import com.pranav.common.util.FileUtil;
import com.pranav.lib_android.exception.CompilationFailedException;
import com.pranav.lib_android.interfaces.*;

import org.eclipse.jdt.internal.compiler.batch.Main;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.util.ArrayList;

public class ECJCompilationTask extends Task {

    private final StringBuilder errs = new StringBuilder();
    private final SharedPreferences prefs;

    public ECJCompilationTask(Builder builder) {
        prefs =
                builder.getContext()
                        .getSharedPreferences("compiler_settings", Context.MODE_PRIVATE);
    }

    @Override
    public String getTaskName() {
        return "ECJ Compilation Task";
    }

    @Override
    public void doFullTask() throws Exception {

        var writer =
                new PrintWriter(
                        new OutputStream() {
                            @Override
                            public void write(int p1) throws IOException {
                                errs.append((char) p1);
                            }
                        });

        var main = new Main(writer, writer, false, null, null);

        var output = new File(FileUtil.getBinDir(), "classes");

        ConcurrentUtil.execute(
                () -> {
                    final var args = new ArrayList<String>();

                    args.add("-log");
                    args.add(FileUtil.getBinDir().concat("debug.xml"));
                    args.add("-g");
                    args.add("-" + prefs.getString("javaVersion", "7.0"));
                    args.add("-d");
                    args.add(output.getAbsolutePath());
                    args.add("-classpath");
                    args.add(FileUtil.getClasspathDir() + "android.jar");
                    var classpath = new StringBuilder();
                    if (prefs.getString("javaVersion", "7.0").equals("8.0")) {
                        classpath.append(FileUtil.getClasspathDir() + "core-lambda-stubs.jar");
                    }
                    var clspath = prefs.getString("classpath", "");
                    if (!clspath.isEmpty() && classpath.length() > 0) {
                        classpath.append(":");
                        classpath.append(clspath);
                    }
                    if (classpath.length() > 0) {
                        args.add("-cp");
                        args.add(classpath.toString());
                    }
                    args.add("-proc:none");
                    args.add("-sourcepath");
                    args.add(" ");
                    args.add(FileUtil.getJavaDir());

                    main.compile(args.toArray(new String[0]));
                });

        if (main.globalErrorsCount > 0 | !output.exists()) {
            throw new CompilationFailedException(errs.toString());
        }
    }
}

File: android-compiler/src/main/java/org/cosmic/ide/android/task/java/ECJCompilationTask.kt
package org.cosmic.ide.android.task.java

import android.content.SharedPreferences
import org.cosmic.ide.android.exception.CompilationFailedException
import org.cosmic.ide.android.interfaces.Task
import org.cosmic.ide.common.util.FileUtil
import org.cosmic.ide.project.Project
import org.eclipse.jdt.internal.compiler.batch.Main
import java.io.File
import java.io.OutputStream
import java.io.PrintWriter

class ECJCompilationTask(preferences: SharedPreferences) : Task {

    private val errs = StringBuilder()
    private val prefs: SharedPreferences

    init {
        prefs = preferences
    }

    override fun getTaskName(): String {
        return "ECJ Compilation Task"
    }

    @Throws(Exception::class)
    override fun doFullTask(project: Project) {

        val writer =
            PrintWriter(
                object : OutputStream() {
                    override fun write(p1: Int) {
                        errs.append(p1.toChar())
                    }
                })

        val main = Main(writer, writer, false, null, null)

        val output = File(project.getBinDirPath(), "classes")

        val classpath = StringBuilder()
        classpath.append(FileUtil.getClasspathDir() + "android.jar")
        classpath.append(File.pathSeparator)
        classpath.append(FileUtil.getClasspathDir() + "core-lambda-stubs.jar")
        classpath.append(File.pathSeparator)
        classpath.append(FileUtil.getClasspathDir() + "kotlin-stdlib-1.7.20-RC.jar")
        classpath.append(File.pathSeparator)
        classpath.append(FileUtil.getClasspathDir() + "kotlin-stdlib-common-1.7.20-RC.jar")
        classpath.append(File.pathSeparator)
        classpath.append(output)
        val clspath = prefs.getString("key_classpath", "")
        if (!clspath!!.isEmpty() && classpath.length > 0) {
            classpath.append(File.pathSeparator)
            classpath.append(clspath)
        }
        val libs = File(project.getLibDirPath()).listFiles()
        if (libs != null) {
            for (lib in libs) {
                classpath.append(File.pathSeparator)
                classpath.append(lib.absolutePath)
            }
        }

        val args = arrayListOf<String>(
            "-g",
            "-" + prefs.getString("key_java_version", "7"),
            "-d",
            output.getAbsolutePath(),
            "-cp",
            classpath.toString(),
            "-proc:none",
            "-sourcepath",
            " ", // This space is needed here
            project.getSrcDirPath()
        )

        main.compile(args.toTypedArray())

        if (main.globalErrorsCount > 0 || !output.exists()) {
            throw CompilationFailedException(errs.toString())
        }
    }
}
